README.md
# http

基于axios封装的请求库

## todo

- [x] 换成ts
- [ ] npm自动发布


CHANGELOG.md
## [1.1.1](https://github.com/youtingkun/http/compare/v1.1.0...v1.1.1) (2022-03-28)


### Features

* 修改 ([07c161b](https://github.com/youtingkun/http/commit/07c161bb2bccc399aa15f98c1ec55660599b8130))



# [1.1.0](https://github.com/youtingkun/http/compare/v1.0.9...v1.1.0) (2022-03-25)


### Features

* name ([f84743f](https://github.com/youtingkun/http/commit/f84743f3b6ee5f2e3e4dec9bfb9cc5d82a0ccdea))



## [1.0.9](https://github.com/youtingkun/http/compare/v1.0.8...v1.0.9) (2022-03-25)


### Features

* name ([e57fa4e](https://github.com/youtingkun/http/commit/e57fa4e80d7e7ac04c9fc4709b9ec1551e98cc3a))



## [1.0.8](https://github.com/youtingkun/http/compare/v1.0.7...v1.0.8) (2022-03-25)


### Features

* name ([a0b646f](https://github.com/youtingkun/http/commit/a0b646f9f177c79ced92c5c2e4cb99b656dd38ae))



## [1.0.7](https://github.com/youtingkun/http/compare/v1.0.6...v1.0.7) (2022-03-25)


### Features

* name ([87cbcb2](https://github.com/youtingkun/http/commit/87cbcb2e49482524592a9c3b6bf2666e3900520a))



## [1.0.6](https://github.com/youtingkun/http/compare/v1.0.5...v1.0.6) (2022-03-25)


### Features

* name ([fbb0ba9](https://github.com/youtingkun/http/commit/fbb0ba9583d076e5ba4a122c03550b1845d7c7ab))



## [1.0.5](https://github.com/youtingkun/http/compare/v1.0.4...v1.0.5) (2022-03-25)


### Features

* name ([1258233](https://github.com/youtingkun/http/commit/1258233f9ec3af583a39d3a87bab2170ebc06200))



## [1.0.4](https://github.com/youtingkun/http/compare/v1.0.3...v1.0.4) (2022-03-25)


### Features

* ts ([635a4d8](https://github.com/youtingkun/http/commit/635a4d81f0fa345ed851c9825e8b7cdb58e6f1a1))



## [1.0.3](https://github.com/youtingkun/http/compare/v1.0.2...v1.0.3) (2022-03-25)



## 1.0.2 (2022-03-25)


### Features

* 发布 ([81ce4f7](https://github.com/youtingkun/http/commit/81ce4f728a42de4782868080b30e84e005c6c1fb))
* init ([fb79f2d](https://github.com/youtingkun/http/commit/fb79f2d4913b3b0c7c2cc6610fca23173ab540db))
* ts ([1bc4cb1](https://github.com/youtingkun/http/commit/1bc4cb1b951547f13f4d218be826a66ebacb4b4e))
* ts ([61a2294](https://github.com/youtingkun/http/commit/61a2294cd366f506d64d4bedbecdcf80a9b888f5))



## 1.0.1 (2022-03-25)


### Features

* 发布 ([81ce4f7](https://github.com/youtingkun/http/commit/81ce4f728a42de4782868080b30e84e005c6c1fb))
* init ([fb79f2d](https://github.com/youtingkun/http/commit/fb79f2d4913b3b0c7c2cc6610fca23173ab540db))
* ts ([1bc4cb1](https://github.com/youtingkun/http/commit/1bc4cb1b951547f13f4d218be826a66ebacb4b4e))
* ts ([61a2294](https://github.com/youtingkun/http/commit/61a2294cd366f506d64d4bedbecdcf80a9b888f5))





index.js

import axios, { AxiosRequestConfig, AxiosResponse } from "axios";

const showStatus = (status) => {
  let mes = "";
  switch (status) {
    case 400:
      mes = "请求错误(400)";
      break;
    case 401:
      mes = "未授权，请重新登录(401)";
      break;
    case 403:
      mes = "拒绝访问(403)";
      break;
    case 404:
      mes = "请求出错(404)";
      break;
    case 408:
      mes = "请求超时(408)";
      break;
    case 500:
      mes = "服务器错误(500)";
      break;
    case 501:
      mes = "服务未实现(501)";
      break;
    case 502:
      mes = "网络错误(502)";
      break;
    case 503:
      mes = "服务不可用(503)";
      break;
    case 504:
      mes = "网络超时(504)";
      break;
    case 505:
      mes = "HTTP版本不受支持(505)";
      break;
    default:
      mes = `连接出错(${status})!`;
  }
  return `${mes}，请检查网络或联系管理员！`;
};

const service = axios.create({
  // 判断环境
  baseURL: process.env.NODE_ENV === "production" ? `/` : "/apis",
  headers: {
    get: {
      "Content-Type": "application/x-www-form-urlencoded;charset=utf-8"
    },
    post: {
      "Content-Type": "application/json;charset=utf-8"
    }
  },
  // 是否跨站点访问控制请求
  withCredentials: true,
  timeout: 30000,
  transformRequest: [
    data => {
      data = JSON.stringify(data);
      return data;
    }
  ],
  validateStatus() {
    // 使用async-await，处理reject情况较为繁琐，所以全部返回resolve，在业务代码中处理异常
    return true;
  },
  transformResponse: [
    data => {
      if (typeof data === "string" && data.startsWith("{")) {
        data = JSON.parse(data);
      }
      return data;
    }
  ]
});

// 请求拦截器
service.interceptors.request.use(
  (config) => {
    return config;
  },
  error => {
    // 错误抛到业务代码
    error.data = {};
    error.data.msg = "服务器异常，请联系管理员！";
    return Promise.resolve(error);
  }
);

// 响应拦截器
service.interceptors.response.use(
  (response) => {
    const status = response.status;
    let msg = "";
    if (status < 200 || status >= 300) {
      // 处理http错误，抛到业务代码,这里的状态码可以为data里面的业务具体的状态码
      msg = showStatus(status);
      if (typeof response.data === "string") {
        response.data = { msg };
      } else {
        response.data.msg = msg;
      }
    }
    return response;
  },
  error => {
    // 错误抛到业务代码，这里的状态码为http所规定的状态码
    error.data = {};
    error.data.msg = "请求超时或服务器异常，请检查网络或联系管理员！";
    return Promise.resolve(error);
  }
);

export default service;


webpack.config.js
const path = require("path");
module.exports = {
  entry: "./src/index.ts",
  output: {
    path: path.resolve(__dirname, "dist"),
    filename: "index.js",
    libraryTarget: "umd", // 保持原函数名
  },
  resolve: {
    extensions: [".ts", ".js"],
  },
  module: {
    // 处理ts或tsx文件的两种方案：
    // 方案1：对ts文件也使用babel-loader，然后在.babelrc里面使用"@babel/preset-typescript"，不会使用tsconfig.json里面的配置。
    // 方案2：对ts文件使用ts-loader，先把ts/tsx 转成 js/jsx，会使用tsconfig.json里面的配置，然后再用 babel-loader 去调用 babel 系列插件，编译成最终的 js
    rules: [
      //   {
      //     // 同时匹配js或者ts文件
      //     test: /\.(js|ts)$/,
      //     exclude: /node_modules/,
      //     use: [
      //       {
      //         loader: "babel-loader",
      //       },
      //     ],
      //   },

      {
        test: /\.ts$/,
        exclude: /node_modules/,
        use: ["babel-loader", { loader: "ts-loader" }], // 执行顺序从右到左
      },
    ],
  },
  mode: "production",
};


src/index.ts
import axios, { AxiosRequestConfig, AxiosResponse, AxiosInstance } from "axios";

const showStatus = (status: Number) => {
  let mes = "";
  switch (status) {
    case 400:
      mes = "请求错误(400)";
      break;
    case 401:
      mes = "未授权，请重新登录(401)";
      break;
    case 403:
      mes = "拒绝访问(403)";
      break;
    case 404:
      mes = "请求出错(404)";
      break;
    case 408:
      mes = "请求超时(408)";
      break;
    case 500:
      mes = "服务器错误(500)";
      break;
    case 501:
      mes = "服务未实现(501)";
      break;
    case 502:
      mes = "网络错误(502)";
      break;
    case 503:
      mes = "服务不可用(503)";
      break;
    case 504:
      mes = "网络超时(504)";
      break;
    case 505:
      mes = "HTTP版本不受支持(505)";
      break;
    default:
      mes = `连接出错(${status})!`;
  }
  return `${mes}，请检查网络或联系管理员！`;
};

// 默认配置
const service: AxiosInstance = axios.create({
  // 判断环境
  baseURL: process.env.NODE_ENV === "production" ? `/` : "/apis",
  headers: {
    // get: {
    //   "Content-Type": "application/x-www-form-urlencoded;charset=utf-8"
    // },
    // post: {
    //   "Content-Type": "application/json;charset=utf-8"
    // }
  },
  // 是否跨站点访问控制请求
  withCredentials: true,
  // 请求超时时间
  timeout: 30000,
  // 在请求前对数据进行处理
  transformRequest: [
    (data) => {
      data = JSON.stringify(data);
      return data;
    },
  ],
  // 如果 `validateStatus` 返回 `true` (或设置为 `null` 或 `undefined`)，promise 将被 resolve; 否则，promise 将被 reject。
  validateStatus() {
    // 使用async-await，处理reject情况较为繁琐，所以全部返回resolve，在业务代码中处理异常
    return true;
  },
  // 在请求后对数据进行处理
  transformResponse: [
    (data) => {
      if (typeof data === "string" && data.startsWith("{")) {
        data = JSON.parse(data);
      }
      return data;
    },
  ],
});

// 请求拦截器
service.interceptors.request.use(
  (config: AxiosRequestConfig) => {
    return config;
  },
  (error) => {
    // 错误抛到业务代码
    error.data = {};
    error.data.msg = "服务器异常，请联系管理员！";
    return Promise.resolve(error);
  }
);

// 响应拦截器
service.interceptors.response.use(
  (response: AxiosResponse) => {
    const status = response.status;
    let msg = "";
    if (status < 200 || status >= 300) {
      // 处理http错误，抛到业务代码,这里的状态码可以为data里面的业务具体的状态码
      msg = showStatus(status);
      if (typeof response.data === "string") {
        response.data = { msg };
      } else {
        response.data.msg = msg;
      }
    }
    return response;
  },
  (error) => {
    // 错误抛到业务代码，这里的状态码为http所规定的状态码
    error.data = {};
    error.data.msg = "请求超时或服务器异常，请检查网络或联系管理员！";
    return Promise.resolve(error);
  }
);

export default service;
